File: java/알고리즘별/Backtracking/15657(N과_M_(8)).java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    static int N;
    static int M;
    static ArrayList<Integer> arr = new ArrayList<>();
    static ArrayList<Integer> answer = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        StringTokenizer st1 = new StringTokenizer(br.readLine(), " ");

        N = Integer.parseInt(st1.nextToken());
        M = Integer.parseInt(st1.nextToken());

        StringTokenizer st2 = new StringTokenizer(br.readLine(), " ");

        for (int i = 0; i < N; i++) {
            arr.add(Integer.parseInt(st2.nextToken()));
        }

        Collections.sort(arr);

        Backtracking(0, 0);

        br.close();
        bw.close();
    }

    // IOException을 throws해주어야 bw.write가 가능하다.
    public static void Backtracking(int depth, int index) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        if (depth == M) {
            for (int i = 0; i < M; i++) {
                bw.write(Integer.toString(answer.get(i))+ " ");
            }

            bw.newLine();
            bw.flush();

            return;
        }

        for (int i = index; i < N; i++) {
            answer.add(arr.get(i));
            Backtracking(depth + 1, i);
            answer.remove(answer.size() - 1); // 리스트에서 마지막 요소를 제거하여 인덱스를 유지한다. (pop_back과 동일한 역할)
        }
    }
}

File: cpp/알고리즘별/Backtracking/15657(N과_M_(8)).cpp
// 중복조합과 관련된 문제

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int N, M;
vector <int> v;
vector <int> answer;

void Backtracking(int depth, int idx);

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> N >> M;

    v.resize(N);

    for (int i = 0; i < N; i++) {
        cin >> v[i];
    }

    sort(v.begin(), v.end());

    Backtracking(0, 0);

    return 0;
}

void Backtracking(int depth, int idx) {
    if (depth == M) {
        for (int i = 0; i < M; i++) {
            cout << answer[i] << " ";
        }

        cout << '\n';

        return;
    }

    // 중복조합이므로 방문 배열이 필요 없고, 중복적으로 숫자가 들어가도록 해야한다.
    for (int i = idx; i < N; i++) {
        answer.push_back(v[i]);
        Backtracking(depth + 1, i);
        answer.pop_back();
    }
}